File: java/Palette.java
public class Palette {

	private Color[] _cache;
	private Image _image;
	private int _redLevel;
	private int _blueLevel;
	private Color insideImageColor;
	private Color scaleImageColor;
	private double factor;

    public Palette(int maxIterations, Image image, int redLevel, int blueLevel) {
		this._image = image;
		this._redLevel = redLevel;
		this._blueLevel = blueLevel;
		this.insideImageColor = this._image.allocateColor(0, 0, 0);
		this.scaleImageColor = this._image.allocateColor(0, 0, 0);
		this.factor = 1 / Math.log(maxIterations) * 255;
		this._cache = new Color[maxIterations];
	}

    public Color getInsideColor() {
		return this.insideImageColor;
    }

    public Color getScaleColor() {
		return this.scaleImageColor;
    }

    public Color getColor(int offset) {
	    if (this._cache[offset] instanceof Color) {
	     	return this._cache[offset];
	    }
		int green = (int)(Math.log(offset) * this.factor);
		Color ret = this._image.allocateColor(this._redLevel, green, this._blueLevel);
		this._cache[offset] = ret;
		return ret;
    }
}

File: python/palette.py
import math

class Palette:
	_cache=0
	_image=0
	_redLevel=0
	_blueLevel=0
	insideImageColor=0
	scaleImageColor=0
	factor=0

	def __init__(self, maxIterations, image, redLevel, blueLevel):
		self._image = image
		self._redLevel = redLevel
		self._blueLevel = blueLevel
		self.insideImageColor = self._image.allocateColor(0, 0, 0)
		self.scaleImageColor = self._image.allocateColor(0, 0, 0)
		self.factor = 1 / math.log(maxIterations) * 255
		self._cache = [None for i in range(maxIterations)]

	def getInsideColor(self):
		return self.insideImageColor

	def getScaleColor(self):
		return self.scaleImageColor

	def getColor(self, offset):
		if self._cache[offset] is not None:
			return self._cache[offset]
		green = int(math.log(offset) * self.factor)
		ret = self._image.allocateColor(self._redLevel, green, self._blueLevel)
		self._cache[offset] = ret
		return ret;
